[Refactor] Phase cleanup part 2

DayKev opened the pull request:
![image](https://github.com/user-attachments/assets/363f7dac-20bc-4b2f-a40d-72ecf53eaf4d)
![image](https://github.com/user-attachments/assets/26a95d1a-bd5a-436f-a178-a3d4bd46097b)

## What are the changes the user will see?
N/A

## Why am I making these changes?
Code modernization.

## What are the changes from a developer perspective?
- Many tsdocs added or updated
- Empty constructors removed
- `public`/etc added to all methods
- Return types added to all methods
- `getField()` added to `PartyMemberPokemonPhase` to replace the common pattern of `this.isPlayer ? globalScene.getPlayerField() : globalScene.getEnemyField()`
- Unused parameter removed from `AttemptCapturePhase.failCatch()`
- `readonly` added to all fields it can be applied to
- Some methods and fields renamed for improved clarity/etc
- Unused phase `EnemyPartyMemberPokemonPhase` removed
- Prepended `abstract-` to the filenames of abstract phases
- `HidePartyExpBarPhase` removed and its functionality was extracted to where it was formerly pushed
- `SummonPhase.getTrainerSlot()` added to replace the common pattern of `!(this.fieldIndex % 2) ? TrainerSlot.TRAINER : TrainerSlot.TRAINER_PARTNER`
- Resolved a bunch of leftover `strict-null` `TODO`s

## How to test the changes?
`npm run test`

## Checklist
- [x] **I'm using `beta` as my base branch**
- [x] There is no overlap with another PR?
- [ ] The PR is self-contained and cannot be split into smaller PRs?
  - Uhh.... oops
- [x] Have I provided a clear explanation of the changes?
- [x] Have I tested the changes manually?
- [x] Are all unit tests still passing? (`npm run test`)

DayKev committed:
Clean up phases

- Replace `integer` with `number`, `IntegerHolder` with `Numberholder`

- Clean up imports:

Use `type` when possible, replace relative imports with absolute,
replace `* from Utils` with named imports

- Add `public`/etc and return types to functions

- Add/update tsdocs and comments

- Clean up some leftover `strict-null` `TODO`s

- Improve typing of some parameters

- Improve formatting for readability

DayKev committed:
Merge branch 'beta' into phase-cleanup-part-2

DayKev committed:
Run prettier and eslint

DayKev committed:
Fix merge issues

DayKev committed:
Organize imports

DayKev committed:
Update more phases

Tempo-anon reviewed (commented):
Can you add comments on the class declaration of each phase while you're here? I notice a lot of them aren't commented

DayKev committed:
Merge branch 'beta' into refactor/phase-cleanup-2

DayKev committed:
More phases cleanup

DayKev committed:
Merge branch 'beta' into refactor/phase-cleanup-2

DayKev committed:
Continue through `EnemyCommandPhase`

DayKev committed:
Continue through `LevelUpPhase`

DayKev committed:
More updates

DayKev committed:
Merge branch 'beta' into refactor/phase-cleanup-2

DayKev committed:
Resolve typecheck/merge issues

DayKev committed:
More changes

DayKev committed:
Split ME phases into their own files

DayKev committed:
Mystery Encounter phases

DayKev committed:
More changes

DayKev committed:
Resolve circular import issue

DayKev committed:
Add a bunch of tsdocs

DayKev committed:
Merge branch 'beta' into refactor/phase-cleanup-2

DayKev committed:
Remove `fixedInt`, fix tsdocs

DayKev committed:
Fix typo in comment

DayKev committed:
Fix merge issue

DayKev committed:
Fix locales

DayKev committed:
Apply `readonly` where applicable, remove empty constructors

DayKev committed:
More updates

DayKev committed:
More updates

DayKev committed:
Organize imports pass

DayKev committed:
More updates

DayKev committed:
Nearly done

DayKev committed:
Merge branch 'beta' into refactor/phase-cleanup-2

DayKev committed:
Resolve merge issues

DayKev committed:
Fix locales submodule

DayKev committed:
Rename abstract phase files with `abstract-`

DayKev committed:
Update `StatStageChangePhase`

DayKev committed:
Add `@extends` to existing tsdocs

DayKev committed:
Add missing renamed file

DayKev committed:
Update tsdocs for Mystery Encounter phases

DayKev committed:
Replace optional params in `PokemonHealPhase` with options interface

DayKev committed:
Replace optional params in `SelectModifierPhase` with options interface

DayKev committed:
Update tests using `new SelectModifierPhase()`

DayKev committed:
Remove leftover commented code in `SelectModifierPhase`

DayKev committed:
Update `EncounterPhase`

MokaStitcher committed:
fix daily mode initialization

MokaStitcher committed:
fix status effects not playing their message

MokaStitcher committed:
fix switch phase getting wrong side of the field

MokaStitcher committed:
fix ME initialization

MokaStitcher committed:
Merge branch 'beta' into refactor/phase-cleanup-2

MokaStitcher committed:
fix game over phase ending too early

DayKev committed:
Remove shortening consts for objects that get modified

DayKev committed:
Merge branch 'beta' into refactor/phase-cleanup-2

DayKev committed:
Resolve merge issues

DayKev committed:
Re-apply organize imports

DayKev committed:
Update `TitlePhase`

DayKev committed:
Fix `TitlePhase` and `TurnStartPhase`

DayKev committed:
Remove unreachable code from `SelectTargetPhase`

DayKev committed:
Replace repeat code with a function in `CommandPhase`

DayKev committed:
Missed a `?`

DayKev committed:
Add `console.warn()` for error scenario in `CommandPhase`

DayKev committed:
Possibly misunderstood the issue with semi-invul and captures

DayKev committed:
Re-use this message instead

DayKev committed:
Merge branch 'beta' into refactor/phase-cleanup-2

DayKev committed:
Resolve merge issues

DayKev committed:
`BattleScene.getField()` no longer returns `null`s

MokaStitcher reviewed (commented):
only a third of the way through, but since i'm stopping and found a few issues, sharing them right now

Also before I forget, I think we can change the `phases/mystery-encounter-phases` folder to just `phases/mystery-encounter`

DayKev committed:
Expand use of `failCatch()` and rename to `failCatchRun()`

Adjust pokeball `else if` to account for pokeball changes

Add `break` statements to `Command.RUN` `if/else if` statements
to account for code flow changes

Co-authored-by: Moka <millennium.stitcher@gmail.com>

MokaStitcher reviewed (commented):
i think i'll want to do some amount of playtesting before approving, but code review is done at least

DayKev committed:
Use `.then()` when hiding party exp bar to match previous code style

DayKev committed:
Apply suggestions from code review

Co-authored-by: Moka <54149968+MokaStitcher@users.noreply.github.com>

DayKev committed:
Rename variables in `SwitchSummonPhase`

DayKev committed:
Remove redundant `.isPlayer()` check in `FaintPhase`

DayKev committed:
Update `CommandPhase.handleCommand()` tsdoc

DayKev committed:
Add `@overload` to `CommandPhase.handleCommand()` tsdocs

DayKev committed:
Replace `any[]` with `unknown[]` in `handleCommand()`

Co-authored-by: innerthunder <168692175+innerthunder@users.noreply.github.com>

DayKev committed:
Rename various `getParty()`-like methods to `getAlliedParty()`/etc

DayKev committed:
Merge branch 'beta' into refactor/phase-cleanup-2

DayKev committed:
Re-add `fixedInt()` for now

This is what is used to make `delayedCall()`s
ignore the game speed multiplier

MokaStitcher reviewed (commented):
Did some playtesting, found one issue. I confirmed that it doesn't happen on current beta. I have not attempted to figure out where it's coming from though

<details><summary>When starting a new daily run the wrong biome background is used</summary>

 - start daily run. Today's starts in metropolis (for the next 3 hours), but the background and platforms are from Town
 - after clearing a wave, next wave's pokemon comes in with the metropolis platform

This happens regardless of which biome is used on the title screen. But when loading runs things are fine.
I'm actually not sure it's purely related to dailies, but since all other modes start in town and town seems to be used as a default i can't get this issue to happen in other modes

![Screenshot_20250101-203958](https://github.com/user-attachments/assets/d84b834c-8534-4f04-b9bb-5b03c62f98da)


</details>

DayKev committed:
Fix typo

Co-authored-by: Moka <54149968+MokaStitcher@users.noreply.github.com>

MokaStitcher reviewed (dismissed):
once that fix gets in it'll be good on my end

<details><summary>much better</summary>

(today's starting biome is Plains so i changed it to something easier to see)

![image](https://github.com/user-attachments/assets/2101c1de-9e36-476a-a49e-82e84b32d9c7)

</details>

DayKev committed:
Fix biome background display issue

Co-authored-by: Moka <54149968+MokaStitcher@users.noreply.github.com>

DayKev committed:
Fix Safari Zone Encounter ME

DayKev committed:
Move tsdoc imports to the top

DayKev committed:
Change `ModifierRewardPhase` to extend `Phase`

DayKev committed:
Add/update tsdocs

DayKev committed:
Fix type error

DayKev committed:
Merge branch 'beta' into refactor/phase-cleanup-2

torranx commented on src/battle.ts:
```diff
@@ -179,8 +179,8 @@ export default class Battle {
-          //@ts-ignore - this is awful to fix/change
-          ret.pokemonId = null;
+          // this is awful to fix/change
+          ret.pokemonId = null!;
```
why is there a bang for `null` here?

DayKev replied:
The bang replaces the `@ts-ignore`. It achieves a similar effect (telling the compiler to allow the `null`) without disabling *every* error.

torranx replied:
Isn't using `@ts-ignore` better than using bang on a `null`? (sorry it's just looks too off to me because we're telling the compiler it's not null but it is actually null)

DayKev replied:
`@ts-ignore` will disable all errors from being displayed, but we only want to disable the error from the `null` assignment. I already ran into this causing a problem once before, an assignment to a `readonly` field was hidden by a `// @ts-ignore` for a different error.

`!` is basically a more limited error suppression command.

`@ts-ignore` should only be used as a last resort imo.

torranx replied:
ok but can we comment the reason why this is `null` and probably a `TODO` as well so that someone who'll look at this won't get confused?

DayKev replied:
I have no idea what the reason is but I can add `TODO` to it.

torranx commented on src/data/move-attrs/hit-heal-attr.ts:
```diff
@@ -60,7 +60,13 @@ export class HitHealAttr extends MoveEffectAttr {
-    globalScene.unshiftPhase(new PokemonHealPhase(user.getBattlerIndex(), healAmount, message, false, true));
+    globalScene.unshiftPhase(
+      new PokemonHealPhase(user.getBattlerIndex(), healAmount, {
+        message: message,
```
```suggestion
        message,
```

torranx commented on src/phases/attempt-run-phase.ts:
```diff
@@ -63,7 +66,13 @@ export class AttemptRunPhase extends PokemonPhase {
-  attemptRunAway(playerField: PlayerPokemon[], enemyField: EnemyPokemon[], escapeChance: NumberHolder) {
+  /**
+   * Calculates the base escape chance based on the current field.
+   * @param playerField - Usually {@linkcode BattleScene.getPlayerField()}
+   * @param enemyField - Usually {@linkcode BattleScene.getEnemyField()}
```
Why is this termed "usually"? Also I suggest removing the `()`

DayKev replied:
Rewrote the param docs.

torranx reviewed (commented):
46/153 files reviewed

DayKev committed:
Update tsdocs in `AttemptRunPhase`

DayKev committed:
Replace `message: message` with `message`

Co-authored-by: Adrian T. <68144167+torranx@users.noreply.github.com>

DayKev committed:
Update comment in `battle.ts`

MokaStitcher reviewed (approved):
played a few more waves for good measure. Time to send it o7
